yourbench/pipeline/ingestion.py

```diff
@@ -0,0 +1,298 @@
+# ingestion.py
+
+"""
+Author: @sumukshashidhar
+
+This module implements the "ingestion" stage of the YourBench pipeline.
+
+Purpose:
+    The ingestion stage reads source documents from a user-specified directory,
+    converts each document into markdown (optionally assisted by an LLM), and
+    saves the converted outputs in the specified output directory. This normalized
+    markdown output sets the foundation for subsequent pipeline steps.
+
+Usage:
+    from yourbench.pipeline import ingestion
+    ingestion.run(config)
+
+Configuration Requirements (in `config["pipeline"]["ingestion"]`):
+    {
+      "run": bool,  # Whether to enable the ingestion stage
+      "source_documents_dir": str,  # Directory containing raw source documents
+      "output_dir": str,           # Directory where converted .md files are saved
+    }
+
+    Additionally, LLM details can be defined in:
+    config["model_roles"]["ingestion"] = [list_of_model_names_for_ingestion]
+    config["model_list"] = [
+      {
+        "model_name": str,
+        "request_style": str,
+        "base_url": str,
+        "api_key": str,
+        ...
+      },
+      ...
+    ]
+
+Stage-Specific Logging:
+    All major ingestion activity is logged to "logs/ingestion.log".
+"""
+
+import glob
+import os
+from dataclasses import dataclass, field
+from typing import Any, Optional
+from huggingface_hub import InferenceClient
+from loguru import logger
+from markitdown import MarkItDown
+
+
+@dataclass
+class IngestionConfig:
+    """Configuration for the ingestion stage of the pipeline."""
+    run: bool = False
+    source_documents_dir: Optional[str] = None
+    output_dir: Optional[str] = None
+
+
+@dataclass
+class ModelConfig:
+    """Configuration for a model in the model list."""
+    model_name: str = "unknown_model"
+    request_style: str = ""
+    base_url: str = ""
+    api_key: str = ""
+
+
+@dataclass
+class ModelRoles:
+    """Configuration for model roles in the pipeline."""
+    ingestion: list[str] = field(default_factory=list)
+
+
+@dataclass
+class PipelineConfig:
+    """Main configuration for the pipeline."""
+    pipeline: dict[str, Any] = field(default_factory=dict)
+    model_roles: ModelRoles = field(default_factory=ModelRoles)
+    model_list: list[ModelConfig] = field(default_factory=list)
+
+
+def _extract_ingestion_config(config: dict[str, Any]) -> IngestionConfig:
+    """
+    Extract ingestion configuration from the main config dictionary.
+    
+    Args:
+        config (dict[str, Any]): The complete configuration dictionary.
```

Comment:
again config will have to change to a proper system, let's open an issue for this